rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/EvaluatorUtil.java

```diff
@@ -93,23 +94,42 @@ public class EvaluatorUtil {
     public static <T> InterruptibleFuture<T> runEvaluator(String task, CompletableFuture<Evaluator> eval, Function<Evaluator, T> call, T defaultResult, Executor exec, boolean isParametric, LanguageClient client) {
         AtomicBoolean interrupted = new AtomicBoolean(false);
         AtomicReference<@Nullable Evaluator> runningEvaluator = new AtomicReference<>(null);
-        return new InterruptibleFuture<>(eval.thenApplyAsync(actualEval -> {
+        AtomicReference<InterruptibleFuture<T>> future = new AtomicReference<>();
+
+        future.set(new InterruptibleFuture<>(eval.thenApplyAsync(actualEval -> {
             try {
+                while (future.get() == null) {
```

Comment:
to be completly threadsafe, we should store it in a local variable and check that against null.

For example:

```
InterruptibleFuture<T> ourself;
while((ourself = future.get()) == null) {
   Thread.yield();
}
```
and use `ourself`  afterwards and know for a fact it cannot be null.